web/pdf_find_controller.js

```diff
@@ -86,20 +86,55 @@ const SPECIAL_CHARS_REG_EXP =
 const NOT_DIACRITIC_FROM_END_REG_EXP = /([^\p{M}])\p{M}*$/u;
 const NOT_DIACRITIC_FROM_START_REG_EXP = /^\p{M}*([^\p{M}])/u;
 
-let normalizationRegex = null;
+// The range [AC00-D7AF] corresponds to the Hangul syllables.
+// The few other chars are some CJK Compatibility Ideographs.
+const SYLLABLES_REG_EXP = /[\uAC00-\uD7AF\uFA6C\uFACF-\uFAD1\uFAD5-\uFAD7]+/g;
+const SYLLABLES_LENGTHS = new Map(
+  [
+    [...Array(0xd7af - 0xac00 + 1).keys()].map(x => x + 0xac00),
+    0xfa6c,
+    [...Array(0xfad1 - 0xfacf + 1).keys()].map(x => x + 0xfacf),
+    [...Array(0xfad7 - 0xfad5 + 1).keys()].map(x => x + 0xfad5),
+  ]
+    .flat()
+    .map(x => {
+      x = String.fromCharCode(x);
+      return [x, x.normalize("NFD").length];
+    })
+);
+// When decomposed (in using NFD) the above syllables will start
+// with one of the chars in this regexp.
+const FIRST_CHAR_SYLLABLES_REG_EXP =
+  "[\\u1100-\\u1112\\ud7a4-\\ud7af\\ud84a\\ud84c\\ud850\\ud854\\ud857\\ud85f]";
+
 function normalize(text) {
   // The diacritics in the text or in the query can be composed or not.
   // So we use a decomposed text using NFD (and the same for the query)
   // in order to be sure that diacritics are in the same order.
 
-  if (!normalizationRegex) {
-    // Compile the regular expression for text normalization once.
-    const replace = Object.keys(CHARACTERS_TO_NORMALIZE).join("");
-    normalizationRegex = new RegExp(
-      `([${replace}])|(\\p{M}+(?:-\\n)?)|(\\S-\\n)|(\\n)`,
-      "gum"
-    );
+  // Collect syllables length and positions.
+  const syllablePositions = [];
+  let m;
+  while ((m = SYLLABLES_REG_EXP.exec(text)) !== null) {
+    let { index } = m;
+    for (const char of m[0]) {
+      syllablePositions.push([SYLLABLES_LENGTHS.get(char), index++]);
+    }
+  }
+
+  // Compile the regular expression for text normalization once.
+  const replace = Object.keys(CHARACTERS_TO_NORMALIZE).join("");
+  let regexp = `([${replace}])|(\\p{M}+(?:-\\n)?)|(\\S-\\n)|(\\n)`;
```

Comment:
This caching has now been lost, can you please restore it (since this part is indeed *constant*)?